### PR TITLE
MAM legacy

### PR DESCRIPTION
**Pull request by muXxer**
# Description

This PR adds MAM legacy to iota.go

## Type of change

- [x] Enhancement (non-breaking change which adds functionality)

# How Has This Been Tested?
I added test cases for message creation and parsing.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have written example code according to the contribution guidelines
- [x] I have added tests using ginkgo that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by sobolev-alexey** [comment]
Awesome, thanks @muXxer!

**Comment by muXxer** [comment]
Alexey could you add the missing layer which adds mode public, private, restricted and pushes the messages to the tangle, like you did in JS? :)

File: mam/v1/mam.go [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+//	err is the error message
+func MamCreate(payloadLength uint64,
+	message Trytes, sideKey Trytes,
+	merkleTree Trits, merkleTreeLength uint64,
+	leafCount uint64, index uint64,
+	nextRoot Trits, start uint64,
+	seed Trytes, security SecurityLevel) (payload Trits, payloadMinLength uint64, err error) {
+
+	if security > 3 {
+		return nil, 0, fmt.Errorf("invalid security %v", security)
```
Comment (luca-moser) [inline mam/v1/mam.go]: Use `ValidateSecurityLevel()` from the `validators` package.

File: mam/v1/mam.go [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+		return nil, 0, err
+	}
+
+	signatureLength := uint64(security) * ISSKeyLength
+	payloadMinLength = encIndexLength + encMessageLengthLenght + HashTrinarySize +
+		uint64(messageLength) + HashTrytesSize + signatureLength +
+		encSiblingsNumberLength + (siblingsNumber * HashTrinarySize)
+
+	if payloadLength < payloadMinLength {
+		return nil, 0, fmt.Errorf("payload too short: needed %d, given %d", payloadMinLength, payloadLength)
```
Comment (luca-moser) [inline mam/v1/mam.go]: Make an explicit `Err...` variable at the start of the file and use `errors.Wrapf(Err..., "needed %d, given %d", minLength, payloadLength). So people can actually type switch on the origin error.

File: mam/v1/mam.go [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+	}
+
+	var offset uint64 = 0
```
Comment (luca-moser) [inline mam/v1/mam.go]:
```suggestion
	var offset uint64
```

File: mam/v1/mam.go [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+	"github.com/iotaledger/iota.go/signing/utils"
+	. "github.com/iotaledger/iota.go/trinary"
+)
+
+// MamInitEncryption initializes the encryption/decryption state for a MAM session
+//
+//	sideKey is the encryption/decryption key
+//	merkleRoot is the merkle root
+//	spongeFunc is the spongeFunction instance used for encryption/decryption
+func MamInitEncryption(sideKey Trits, merkleRoot Trits, spongeFunc sponge.SpongeFunction) {
```
Comment (luca-moser) [inline mam/v1/mam.go]: Should be called `MAM...`

File: mam/v1/mam.go [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+	spongeFunc.Absorb(sideKey)
+	spongeFunc.Absorb(merkleRoot[:HashTrinarySize])
+}
+
+// PayloadMinLength computes the minimum length of a payload
```
Comment (luca-moser) [inline mam/v1/mam.go]:
```suggestion
// PayloadMinLength computes the minimum length of a payload.
```

File: mam/v1/mam.go [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+	. "github.com/iotaledger/iota.go/consts"
+	"github.com/iotaledger/iota.go/curl"
+	. "github.com/iotaledger/iota.go/curl/hamming"
+	. "github.com/iotaledger/iota.go/merkle"
+	"github.com/iotaledger/iota.go/signing/legacy"
+	"github.com/iotaledger/iota.go/signing/utils"
+	. "github.com/iotaledger/iota.go/trinary"
+)
+
+// MamInitEncryption initializes the encryption/decryption state for a MAM session
```
Comment (luca-moser) [inline mam/v1/mam.go]:
```suggestion
// MamInitEncryption initializes the encryption/decryption state for a MAM session.
```

File: mam/v1/mam.go [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+//	nextRoot is the merkle root of the next MAM payload
+//	start is the offset used to generate addresses
+//	seed is the seed used to generate addresses - Not sent over the network
+//	security is the security level used to generate addresses
+//
+// Returns:
+//  payload is the payload of the encrypted message
+//  payloadMinLength is the length of the payload needed to encrypt the message
+//	err is the error message
+func MamCreate(payloadLength uint64,
```
Comment (luca-moser) [inline mam/v1/mam.go]: `MAM...`

File: mam/v1/mam.go [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+	message Trytes, sideKey Trytes,
+	merkleTree Trits, merkleTreeLength uint64,
+	leafCount uint64, index uint64,
+	nextRoot Trits, start uint64,
+	seed Trytes, security SecurityLevel) (payload Trits, payloadMinLength uint64, err error) {
```
Comment (luca-moser) [inline mam/v1/mam.go]: Declare the return variables inside the function instead in the function signature.

File: mam/v1/mam.go [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+//	sideKey is the decryption key
+//	root is the merkle root
+//
+// Returns:
+//  parsedIndex is the parsed index of the message
+//  parsedNextRoot is the merkle root of the next MAM payload
+//	parsedMessage is the decrypted message
+//	parsedSecurity is the parsed security level of the message
+//	err is the error message
+func MamParse(payload Trits, payloadLength uint64, sideKey Trytes, root Trits) (parsedIndex uint64, parsedNextRoot Trytes, parsedMessage Trytes, parsedSecurity SecurityLevel, err error) {
```
Comment (luca-moser) [inline mam/v1/mam.go]: `MAM...`

and declare the return variables inside the function.

File: mam/v1/mam.go [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+	if err != nil {
+		return 0, "", "", SecurityLevel(0), err
+	}
+
+	encCurl := curl.NewCurlP27().(*curl.Curl)
+	MamInitEncryption(sideKeyTrits, root, encCurl)
+
+	// decode index from payload
+	if offset >= payloadLength {
+		return 0, "", "", SecurityLevel(0), errors.New("payload too small")
```
Comment (luca-moser) [inline mam/v1/mam.go]: Make a separate `Err...` var for this kind of error.

File: mam/v1/mam.go [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+	index, encIndexLength, err := DecodeInt64(payload[offset:])
+	if err != nil {
+		return 0, "", "", SecurityLevel(0), err
+	}
+	parsedIndex = uint64(index)
+	offset += encIndexLength
+
+	// decode message length from payload
+	if offset >= payloadLength {
+		return 0, "", "", SecurityLevel(0), errors.New("payload too small")
```
Comment (luca-moser) [inline mam/v1/mam.go]: Make a separate `Err...` var for this kind of error.

File: mam/v1/mam.go [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+	}
+
+	messageLength, encMessageLengthLenght, err := DecodeInt64(payload[offset:])
+	if err != nil {
+		return 0, "", "", SecurityLevel(0), err
+	}
+	offset += encMessageLengthLenght
+
+	if offset >= payloadLength {
+		return 0, "", "", SecurityLevel(0), errors.New("payload too small")
```
Comment (luca-moser) [inline mam/v1/mam.go]: Make a separate `Err...` var for this kind of error.

File: mam/v1/mam.go [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+	encCurl.Absorb(payload[:offset])
+
+	// decrypt next root from payload
+	nextRoot := Unmask(payload[offset:], HashTrinarySize, encCurl)
+	parsedNextRoot = MustTritsToTrytes(nextRoot)
+	offset += HashTrinarySize
+
+	// decrypt message from payload
+	if offset >= payloadLength {
+		return 0, "", "", SecurityLevel(0), errors.New("payload too small")
```
Comment (luca-moser) [inline mam/v1/mam.go]: Make a separate `Err...` var for this kind of error.

File: mam/v1/mam.go [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+	}
+	message := Unmask(payload[offset:], uint64(messageLength), encCurl)
+	parsedMessage = MustTritsToTrytes(message)
+
+	offset += uint64(messageLength)
+
+	// decrypt nonce from payload
+	if offset >= payloadLength {
+		return 0, "", "", SecurityLevel(0), errors.New("payload too small")
```
Comment (luca-moser) [inline mam/v1/mam.go]: Make a separate `Err...` var for this kind of error.

File: mam/v1/mam.go [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+	}
+	// Information is returned in encCurl.State
+	Unmask(payload[offset:], uint64(HashTrytesSize), encCurl)
+	offset += HashTrytesSize
+
+	// get security back from state
+	if offset >= payloadLength {
+		return 0, "", "", SecurityLevel(0), errors.New("payload too small")
```
Comment (luca-moser) [inline mam/v1/mam.go]: Make a separate `Err...` var for this kind of error.

File: mam/v1/mam.go [hunk 16]
```diff
@@ -0,0 +1,298 @@
...
+	copy(hash, encCurl.State[:HashTrinarySize])
+
+	parsedSecurity, err = signing.GetSecurityLevel(hash)
+	if err != nil {
+		return 0, "", "", SecurityLevel(0), err
+	}
+	copy(payload[offset:], Unmask(payload[offset:], payloadLength-offset, encCurl))
+
+	if parsedSecurity == 0 {
+		return 0, "", "", SecurityLevel(0), errors.New("wrong security level")
```
Comment (luca-moser) [inline mam/v1/mam.go]: Make a separate `Err...` var for this kind of error.

File: mam/v1/mam.go [hunk 17]
```diff
@@ -0,0 +1,298 @@
...
+	encCurl.Reset()
+	digest, err := signing.Digest(hash, payload[offset:offset+(uint64(parsedSecurity)*ISSKeyLength)], 0, encCurl)
+
+	// complete the address
+	address, err := signing.Address(digest, encCurl)
+	offset += uint64(parsedSecurity) * ISSKeyLength
+
+	// decrypt siblings number from payload
+	if offset >= payloadLength {
+		return 0, "", "", SecurityLevel(0), errors.New("payload too small")
```
Comment (luca-moser) [inline mam/v1/mam.go]: Make a separate `Err...` var for this kind of error.

File: mam/v1/mam.go [hunk 18]
```diff
@@ -0,0 +1,298 @@
...
+	}
+
+	siblingsNumber, encSiblingsNumberLength, err := DecodeInt64(payload[offset:])
+	offset += encSiblingsNumberLength
+
+	// get merkle root from siblings from payload
+	if siblingsNumber != 0 {
+		if offset >= payloadLength {
+			return 0, "", "", SecurityLevel(0), errors.New("payload too small")
```
Comment (luca-moser) [inline mam/v1/mam.go]: Make a separate `Err...` var for this kind of error.

File: mam/v1/mam.go [hunk 19]
```diff
@@ -0,0 +1,298 @@
...
+		}
+		address, err = MerkleRoot(address, payload[offset:], uint64(siblingsNumber), parsedIndex, encCurl)
+	}
+
+	// check merkle root with the given root
+	if !reflect.DeepEqual(address, root[:HashTrinarySize]) {
+		return 0, "", "", SecurityLevel(0), errors.New("Merkle root does not match")
```
Comment (luca-moser) [inline mam/v1/mam.go]: Make a separate `Err...` var for this kind of error.

**Pull request merged**